### PR TITLE
Adapt to Gardener API changes

### LINKED ISSUE #536: Adapt to gardener API Changes
Wait for https://github.com/gardener/gardener-extensions/pull/521 to be released and then adapt to the new fields in the shoot spec.
Also adapt to new gardener API `gardener.cloud/v1beta1`
(see also https://github.com/gardener/gardener/releases/tag/v0.33.0)
Purpose is now part of shoot spec: https://github.com/gardener/gardener/issues/1814

### PR DESCRIPTION
**Pull request by grolu**
**What this PR does / why we need it**:
- Updated to gardener.cloud/v1beta1 API
- Updated to new fields in infra specs
- Use spec.purpose field

**Which issue(s) this PR fixes**:
Fixes #536

**Special notes for your reviewer**:
Screen Shot 2020-02-18 at 12 05 25

**Release note**:

```improvement user
Adapted to `gardener.cloud/v1beta1` API
```
```improvement user
Support for the new `testing` purpose. Shoots with this purpose won't get a logging or monitoring stack as part of their control planes
```

### CODE REVIEW

File: backend/lib/services/shoots.js [hunk 1]
```diff
@@ -81,6 +81,7 @@ exports.replace = async function ({ user, namespace, name, body }) {
   Object.assign(metadata, { labels, annotations })
   // compose new body
   body = { kind, apiVersion, metadata, spec, status }
+  body = { kind, apiVersion, metadata, spec, status }
```
Comment (petersutter) [inline backend/lib/services/shoots.js]: rm

File: backend/test/acceptance/api.shoots.spec.js [hunk 2]
```diff
@@ -66,17 +66,16 @@ module.exports = function ({ agent, sandbox, k8s, auth }) {
...
+        metadata: {
+          name,
+          namespace
+        },
+        spec
+      })
 
     expect(res).to.have.status(200)
     expect(res).to.be.json
-    expect(res.body.metadata).to.eql({ name, namespace, resourceVersion, annotations })
```
Comment (petersutter) [inline backend/test/acceptance/api.shoots.spec.js]: why is this removed?

File: backend/test/acceptance/api.shoots.spec.js [hunk 3]
```diff
@@ -148,8 +147,7 @@ module.exports = function ({ agent, sandbox, k8s, auth }) {
     const bearer = await user.bearer
     const metadata = {
       annotations: {
-        'gardener.cloud/created-by': 'baz@example.org',
-        'garden.sapcloud.io/purpose': 'evaluation'
+        'garden.sapcloud.io/created-by': 'baz@example.org'
```
Comment (petersutter) [inline backend/test/acceptance/api.shoots.spec.js]: something must have gone wrong during merge: gardener.cloud/created-by

File: backend/test/acceptance/api.shoots.spec.js [hunk 4]
```diff
@@ -171,10 +169,7 @@ module.exports = function ({ agent, sandbox, k8s, auth }) {
     const actLabels = body.metadata.labels
     const expLabels = metadata.labels
     expect(actLabels).to.eql(expLabels)
-    const actPurpose = body.metadata.annotations['garden.sapcloud.io/purpose']
-    const expPurpose = metadata.annotations['garden.sapcloud.io/purpose']
-    expect(actPurpose).to.equal(expPurpose)
```
Comment (petersutter) [inline backend/test/acceptance/api.shoots.spec.js]: why is the check removed and not adapted to spec.purpose?

File: frontend/src/components/ShootDetails/ShootJournalsCard.vue [hunk 5]
```diff
@@ -86,7 +89,9 @@ export default {
 
       const createdAt = `**Created At:** ${this.shootCreatedAt}`
       const lastOperation = `**Last Op:** ${get(this.shootLastOperation, 'description', '')}`
-      const lastError = `**Last Error:** ${get(this.shootLastError, 'description', '-')}`
+      let shootLastErrorDescriptions = compact(map(this.shootLastErrors, 'description'))
+      shootLastErrorDescriptions = join(shootLastErrorDescriptions, '\n')
+      const lastError = `**Last Errors:** ${shootLastErrorDescriptions || '-'}`
```
Comment (petersutter) [inline frontend/src/components/ShootDetails/ShootJournalsCard.vue]: have you checked how the endresult looks? Can each error easily be recognized? Or should we use bulleted list / add seperator lines or similar

File: frontend/src/components/ShootStatus.vue [hunk 6]
```diff
@@ -41,14 +41,14 @@ limitations under the License.
       </v-tooltip>
     </div>
     <ansi-text v-if="!!popperMessage" :text="popperMessage"></ansi-text>
-    <template v-if="isError">
-      <v-divider class="my-2"></v-divider>
-      <h4 class="error--text text-xs-left">Last Error</h4>
-      <template v-for="errorCodeDescription in errorCodeDescriptions">
+    <v-divider class="my-2"></v-divider>
+    <h4 class="error--text text-xs-left">Last Errors</h4>
+    <div v-for="lastErrorDescription in lastErrorDescriptions" v-bind:key="lastErrorDescription.description">
```
Comment (petersutter) [inline frontend/src/components/ShootStatus.vue]: as key: use hash? or irrelevant?

File: backend/test/acceptance/api.shoots.spec.js (hunk 2 above)
Comment (grolu) [reply backend/test/acceptance/api.shoots.spec.js]: you are right, probably too eager deleting code...

File: backend/test/acceptance/api.shoots.spec.js (hunk 4 above)
Comment (grolu) [reply backend/test/acceptance/api.shoots.spec.js]: Because as purpose is now part of spec
`expect(body.spec).to.eql(spec)`
should to the trick. We could explicitly check for purpose but we also don't do this for other values.

File: frontend/src/components/ShootDetails/ShootJournalsCard.vue (hunk 5 above)
Comment (grolu) [reply frontend/src/components/ShootDetails/ShootJournalsCard.vue]: I don't know actually. I did not find any shoot with more than one error. I talked to the Gardener Team and they told me that usually there are multiple errors only if simultaneous operations fail, however I could not find such a situation. In general - yes, it would probably make sense to enhance our view here. The general look of this dialog is not very good as we show lengthy error messages. Adding separators could on the one hand improve readability. On the other hand it would also increase required space and could eventually make the dialog too big for the screen (which can happen anyway, depending on the amount and length of the messages). As I did not observe this happening too often, I decided to not invest here atm.

File: frontend/src/components/ShootStatus.vue (hunk 6 above)
Comment (grolu) [reply frontend/src/components/ShootStatus.vue]: We could generate a random value to identify the messages and add it to the errorDescription Object to overcome this. I assumed this to be a hypothetical problem but you are right, it _could_ happen and lead to a warning / out-of-sync dom. I will take care of it.

File: frontend/src/components/ShootStatus.vue [hunk 7]
```diff
@@ -137,22 +137,24 @@ export default {
...
-    },
-    errorCodes () {
-      return get(this.lastError, 'codes', [])
+    lastErrorDescriptions () {
+      return map(this.lastErrors, lastError => {
+        return {
+          description: lastError.description,
+          errorCodeDescriptions: map(lastError.codes, code => get(errorCodes, `${code}.description`, code))
+        }
+      })
```
Comment (petersutter) [inline frontend/src/components/ShootStatus.vue]: return map(this.lastErrors, lastError => ({
        description: lastError.description,
        errorCodeDescriptions: map(lastError.codes, code => get(errorCodes, `${code}.description`, code))
      }))

File: frontend/src/utils/index.js [hunk 8]
```diff
@@ -484,19 +486,19 @@ export function selfTerminationDaysForSecret (secret) {
 }
 
 export function purposesForSecret (secret) {
-  return selfTerminationDaysForSecret(secret) ? ['evaluation'] : ['evaluation', 'development', 'production']
+  return selfTerminationDaysForSecret(secret) ? ['evaluation'] : ['evaluation', 'development', 'production', 'testing']
```
Comment (petersutter) [inline frontend/src/utils/index.js]: we should add descriptions for the purposes (maybe in a create a separate issue?), as e.g. for `testing` purpose the clusters won't have a monitoring and logging stack deployed

Comment (grolu) [reply frontend/src/utils/index.js]: done

**Comment by ialidzhikov** [comment]
I think we have to hide the logging & monitoring tiles when `.spec.purpose=testing`. Is it adapted in this PR?

File: backend/test/acceptance/api.shoots.spec.js (hunk 4 above)
Comment (petersutter) [reply backend/test/acceptance/api.shoots.spec.js]: will you add
`expect(body.spec).to.eql(spec)`?

Comment (grolu) [reply backend/test/acceptance/api.shoots.spec.js]: it is already in

Comment (petersutter) [reply backend/test/acceptance/api.shoots.spec.js]: ah ok, didn't see it in the diff view

**Comment by grolu** [comment]
> I think we have to hide the logging & monitoring tiles when `.spec.purpose=testing`. Is it adapted in this PR?

This is a very good point. I will implement this now

**Comment by petersutter** [comment]
maybe one minor thing: does it make sense to put the purposes in a different order?
- evaluation
- development
- testing
- production

File: frontend/src/components/ShootDetails/ShootMonitoringCard.vue [hunk 9]
```diff
@@ -49,10 +50,13 @@ limitations under the License.
           <status-tags v-else :conditions="shootConditions" popperPlacement="bottom"></status-tags>
         </div>
       </v-card-title>
-      <template v-if="seedShootIngressDomain">
-        <v-divider class="my-2" inset></v-divider>
-        <cluster-metrics :shootItem="shootItem"></cluster-metrics>
-      </template>
+
+      <v-divider class="my-2" inset></v-divider>
+      <v-card-title class="listItem" v-if="!!metricsNotAvailableText">
```
Comment (petersutter) [inline frontend/src/components/ShootDetails/ShootMonitoringCard.vue]:
```suggestion
      <v-card-title class="listItem" v-if="metricsNotAvailableText">
```
